test/metabase/query_processor_test/date_time_zone_functions_test.clj

```diff
@@ -132,3 +142,140 @@
                        :breakout    [[:expression "expr"]]}]]]
               (testing (format "%s function works as expected on %s column for driver %s" op col-type driver/*driver*)
                 (is (= (set expected) (set (test-date-extract query))))))))))))
+
+(defn- date-math
+  [op x amount unit col-type]
+  (let [amount (if (= op :date-add)
+                 amount
+                 (- amount))
+        result (cond
+                 ;; unlike everybody else, vertica decided when adding 1 year it will do that by adding 365 days,
+                 ;; or 90 days if you want to add a quarter.
+                 ;; That means if you do `select date '2004-02-02' + interval '2 year';` in vertica, it will returns
+                 ;; `2006-02-01` because 2004 is a leap year. All others DBs will return `2006-02-02` instead.
+                 (and (= driver/*driver* :vertica) (#{:year :quarter :month} unit))
+                 (u.date/add x :day (* amount (case unit
+                                                :year    365
+                                                :quarter (* 3 30)
+                                                :month   30)))
+                 :else
+                 (u.date/add x unit amount))
+        fmt    (cond
+                 ;; the :date column of :presto should have this format too,
+                 ;; but the test data we created for presto is datetime even if we define it as date
+                 (and (= driver/*driver* :presto) (#{:text-as-date} col-type))
+                 "yyyy-MM-dd"
+
+                 :else
+                 "yyyy-MM-dd HH:mm:ss")]
+    (t/format fmt result)))
+
+(defn- mongo-major-version [db]
+  (-> (get-in db [:details :version])
+      (str/split #"\.")
+      first
+      parse-long))
+
+(deftest date-math-tests
+  (mt/dataset times-mixed
+    (mt/test-drivers (disj (mt/normal-drivers-with-feature :date-functions) :mongo)
+      (doseq [[col-type field-id] [[:datetime (mt/id :times :dt)] [:text-as-datetime (mt/id :times :as_dt)]]]
+        (doseq [op [:date-add :date-subtract]]
+          (doseq [unit [:year :quarter :month :day :hour :minute :second]]
+            (doseq [[expected query]
+                    [[[[(date-math op #t "2004-03-19 09:19:09" 2 unit col-type)] [(date-math op #t "2008-06-20 10:20:10" 2 unit col-type)]
+                       [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)] [(date-math op #t "2012-11-21 11:21:11" 2 unit col-type)]]
```

Comment:
I wonder if it was easier to follow if the field values were read from the database instead of being hardwired here.